File: 数据结构与算法/排序/sort.cs
/*
基于 C# 的排序算法实现
*/

using System.Diagnostics;

internal class Program
{
    private static void Main(string[] args)
    {
        Test(Sort.InsertSort, "InsertSort");
        Test(Sort.ShellSort, "ShellSort");
        Test(Sort.SelectSort, "SelectSort");
        Test(Sort.BubbleSort, "BubbleSort");
        Test(Sort.QuickSort, "QuickSort");
        Test(Sort.HeapSort, "HeapSort");
        Test(Sort.MergeSort, "MergeSort");
        Test(Sort.CountSort, "CountSort");
        Console.WriteLine("All tests passed.");
    }

    private static void Test(Action<IList<int>> sortFunc, string funcName)
    {
        static string str(IEnumerable<int> arr) => string.Join(",", arr);

        var arr = new int[] { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1 };
        sortFunc(arr);
        Trace.Assert(arr.SequenceEqual(Enumerable.Range(1, 10)), $"{funcName}. Expect: {str(Enumerable.Range(1, 10))}. Found: {str(arr)}");

        arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
        sortFunc(arr);
        Trace.Assert(arr.SequenceEqual(Enumerable.Range(1, 10)), $"{funcName}. Expect: {str(Enumerable.Range(1, 10))}. Found: {str(arr)}");

        arr = [1, 10, 2, 9, 3, 8, 4, 7, 5, 6];
        sortFunc(arr);
        Trace.Assert(arr.SequenceEqual(Enumerable.Range(1, 10)), $"{funcName}. Expect: {str(Enumerable.Range(1, 10))}. Found: {str(arr)}");

        arr = [];
        sortFunc(arr);
        Trace.Assert(arr.SequenceEqual(Enumerable.Range(1, 0)), $"{funcName}. Expect: {str(Enumerable.Range(1, 0))}. Found: {str(arr)}");

        arr = [1];
        sortFunc(arr);
        Trace.Assert(arr.SequenceEqual(Enumerable.Range(1, 1)), $"{funcName}. Expect: {str(Enumerable.Range(1, 1))}. Found: {str(arr)}");

        arr = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1];
        sortFunc(arr);
        Trace.Assert(arr.SequenceEqual([1, 1, 1, 1, 1, 1, 1, 1, 1, 1]), $"{funcName}. Expect: {str([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])}. Found: {str(arr)}");

        arr = [1, 1, 2, 2, 3, 3, 4, 4, 5, 5];
        sortFunc(arr);
        Trace.Assert(arr.SequenceEqual([1, 1, 2, 2, 3, 3, 4, 4, 5, 5]), $"{funcName}. Expect: {str([1, 1, 2, 2, 3, 3, 4, 4, 5, 5])}. Found: {str(arr)}");

        arr = [5, 5, 4, 4, 6, 7, 1, 2, 9, 0];
        sortFunc(arr);
        Trace.Assert(arr.SequenceEqual([0, 1, 2, 4, 4, 5, 5, 6, 7, 9]), $"{funcName}. Expect: {str([0, 1, 2, 4, 4, 5, 5, 6, 7, 9])}. Found: {str(arr)}");
    }
}

internal class Sort
{
    // 插入排序
    public static void InsertSort<T>(IList<T> arr) where T : IComparable<T>
    {
        for (int i = 1; i < arr.Count; i++)
            for (int j = i; j > 0; j--)
                if (arr[j].CompareTo(arr[j - 1]) < 0) (arr[j], arr[j - 1]) = (arr[j - 1], arr[j]);
                else break;
    }
    // 希尔排序
    public static void ShellSort<T>(IList<T> arr) where T : IComparable<T>
    {
        for (int gap = arr.Count / 2; gap > 0; gap /= 2)
            for (int i = gap; i < arr.Count; i += gap)
                for (int j = i; j > 0; j -= gap)
                    if (arr[j].CompareTo(arr[j - gap]) < 0) (arr[j], arr[j - gap]) = (arr[j - gap], arr[j]);
                    else break;
    }
    // 选择排序
    public static void SelectSort<T>(IList<T> arr) where T : IComparable<T>
    {
        for (int i = 0; i < arr.Count; i++)
        {
            int minIndex = i;
            for (int j = i + 1; j < arr.Count; j++)
                if (arr[j].CompareTo(arr[minIndex]) < 0) minIndex = j;
            if (minIndex != i) (arr[i], arr[minIndex]) = (arr[minIndex], arr[i]);
        }
    }
    // 冒泡排序
    public static void BubbleSort<T>(IList<T> arr) where T : IComparable<T>
    {
        for (int i = arr.Count - 1; i > 0; i--)
        {
            bool f = false;
            for (int j = 0; j < i; j++)
                if (arr[j].CompareTo(arr[j + 1]) > 0) { (arr[j], arr[j + 1]) = (arr[j + 1], arr[j]); f = true; }
            if (!f) break;
        }
    }
    // 快速排序
    public static void QuickSort<T>(IList<T> arr) where T : IComparable<T>
    {
        void quickSort(int start, int end)
        {
            if (start >= end) return;
            T pivot = arr[start];
            int l = start, r = end;
            while (l < r)
            {
                while (l < r && arr[r].CompareTo(pivot) >= 0) r--;
                if (l < r && arr[r].CompareTo(pivot) < 0) arr[l++] = arr[r];
                while (l < r && arr[l].CompareTo(pivot) <= 0) l++;
                if (l < r && arr[l].CompareTo(pivot) > 0) arr[r--] = arr[l];
            }
            arr[l] = pivot;
            quickSort(start, l - 1);
            quickSort(l + 1, end);
        }
        quickSort(0, arr.Count - 1);
    }
    // 堆排序
    public static void HeapSort<T>(IList<T> arr) where T : IComparable<T>
    {
        void adjust(int index, int end)
        {
            while (true)
            {
                int idx = index;
                if (index * 2 + 1 <= end && arr[index * 2 + 1].CompareTo(arr[idx]) > 0) idx = index * 2 + 1;
                if (index * 2 + 2 <= end && arr[index * 2 + 2].CompareTo(arr[idx]) > 0) idx = index * 2 + 2;
                if (idx == index) break;
                (arr[index], arr[idx]) = (arr[idx], arr[index]);
                index = idx;
            }
        }

        void buildHeap()
        {
            for (int i = (arr.Count - 1 - 1) / 2; i >= 0; i--) adjust(i, arr.Count - 1);
        }

        buildHeap();
        for (int i = arr.Count - 1; i > 0; i--)
        {
            (arr[0], arr[i]) = (arr[i], arr[0]); // 交换
            adjust(0, i - 1);
        }
    }
    // 归并排序
    public static void MergeSort<T>(IList<T> arr) where T : IComparable<T>
    {
        void merge(int l, int m, int r)
        {
            T[] left = new T[m - l];
            for (int idx = l; idx < m; idx++) left[idx - l] = arr[idx];
            T[] right = new T[r - m];
            for (int idx = m; idx < r; idx++) right[idx - m] = arr[idx];
            int i = 0, j = 0, k = l;
            while (i < left.Length && j < right.Length)
            {
                if (left[i].CompareTo(right[j]) <= 0) arr[k++] = left[i++];
                else arr[k++] = right[j++];
            }
            while (i < left.Length) arr[k++] = left[i++];
            while (j < right.Length) arr[k++] = right[j++];
        }

        void mergeSort(int l, int r)
        {
            if (l >= r - 1) return;
            int m = (l + r) / 2;
            mergeSort(l, m);
            mergeSort(m, r);
            merge(l, m, r);
        }

        mergeSort(0, arr.Count);
    }
    // 计数排序
    public static void CountSort(IList<int> arr) // 这个写泛型感觉实在是有点困难...
    {
        if (arr.Count == 0) return;
        int minVal = arr.Min(), maxVal = arr.Max();
        int offset = minVal;
        int[] count = new int[maxVal - minVal + 1];
        foreach (int num in arr) count[num - offset]++;
        int idx = 0;
        for (int i = 0; i < count.Length; i++)
        {
            while (count[idx] == 0) idx++;
            arr[i] = idx + offset;
            count[idx]--;
        }
    }
}

File: DataStructureAndAlgorithm/sort/sort.ts
/*
基于 typescript 实现的常用排序算法
*/

export { };

/**
 * 插入排序
 * @param arr 待排序数组
 */
function insertSort<T>(arr: T[]): void {
    for (let i = 1; i < arr.length; i++) {
        for (let j = i; j > 0; j--) {
            if (arr[j] < arr[j - 1]) {
                [arr[j], arr[j - 1]] = [arr[j - 1], arr[j]]
            }
        }
    }
}

/**
 * 希尔排序
 * @param arr 待排序数组
 */
function shellSort<T>(arr: T[]): void {
    for (let gap = Math.floor(arr.length / 2); gap > 0; gap = Math.floor(gap / 2)) {
        for (let i = gap; i < arr.length; i++) {
            for (let j = i; j >= gap; j -= gap) {
                if (arr[j] < arr[j - gap]) {
                    [arr[j], arr[j - gap]] = [arr[j - gap], arr[j]]
                }
            }
        }
    }
}

/**
 * 选择排序
 * @param arr 待排序数组
 */
function selectSort<T>(arr: T[]): void {
    for (let i = 0; i < arr.length; i++) {
        let minIndex = i
        for (let j = i + 1; j < arr.length; j++) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j
            }
        }
        if (minIndex !== i) {
            [arr[i], arr[minIndex]] = [arr[minIndex], arr[i]]
        }
    }
}

/**
 * 冒泡排序
 * @param arr 待排序数组
 */
function bubbleSort<T>(arr: T[]): void {
    for (let i = arr.length - 1; i >= 0; i--) {
        let f = false;
        for (let j = 0; j < i; j++) {
            if (arr[j] > arr[j + 1]) {
                f = true;
                [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]]
            }
        }
        if (!f) break;
    }
}

/**
 * 快速排序
 * @param arr 待排序数组
 */
function quickSort<T>(arr: T[]): void {
    function quick(left: number, right: number): void {
        if (left >= right) return;
        let i = left, j = right;
        let pivot = arr[left];
        while (i < j) {
            while (i < j && arr[j] >= pivot) j--;
            if (i < j) arr[i++] = arr[j];
            while (i < j && arr[i] < pivot) i++;
            if (i < j) arr[j--] = arr[i];
        }
        arr[i] = pivot;
        quick(left, i - 1);
        quick(i + 1, right);
    }
    quick(0, arr.length - 1);
}

/**
 * 归并排序
 * @param arr 待排序数组
 */
function mergeSort<T>(arr: T[]): void {
    /**
     * 归并两个有序的数组部分
     * @param left 左边开始的索引（闭区间）
     * @param mid 左边结束的索引（开区间）
     * @param right 右边结束的索引（开区间）
     */
    function mergeArr(left: number, mid: number, right: number): void {
        let i = left, j = mid;
        let tmp: T[] = [];
        while (i < mid && j < right) {
            tmp.push(arr[i] < arr[j] ? arr[i++] : arr[j++]);
        }
        while (i < mid) tmp.push(arr[i++]);
        while (j < right) tmp.push(arr[j++]);
        for (let k = 0; k < tmp.length; k++) {
            arr[left + k] = tmp[k];
        }
    }

    /**
     * 归并一个数组区间内的数据
     * @param left 左边开始的索引（闭区间）
     * @param right 右边结束的索引（开区间）
     */
    function merge(left: number, right: number): void {
        if (left >= right - 1) return;
        let mid = (left + right) >> 1;
        merge(left, mid);
        merge(mid, right);
        mergeArr(left, mid, right);
    }

    merge(0, arr.length);
}

/**
 * 堆排序
 * @param arr 待排序数组
 */
function heapSort<T>(arr: T[]): void {
    /**
     * 调整大根堆
     * @param heap 待调整的堆
     * @param start 调整起始位置索引
     * @param end 调整结束位置索引
     */
    function adjust(heap: T[], start: number, end: number): void {
        while (true) {
            let index = start;
            if (start * 2 + 1 <= end && heap[index] < heap[start * 2 + 1]) { index = start * 2 + 1; }
            if (start * 2 + 2 <= end && heap[index] < heap[start * 2 + 2]) { index = start * 2 + 2; }
            if (index === start) break;
            [heap[start], heap[index]] = [heap[index], heap[start]];
            start = index;
        }
    }

    /**
     * 构建大根堆
     * @param arr 待构建为堆的数组
     */
    function buildHeap(arr: T[]): void {
        for (let i = (arr.length - 1 - 1) >> 1; i >= 0; i--) {
            adjust(arr, i, arr.length - 1);
        }
    }

    buildHeap(arr);
    for (let i = arr.length - 1; i > 0; i--) {
        [arr[0], arr[i]] = [arr[i], arr[0]];
        adjust(arr, 0, i - 1);
    }
}

/**
 * 桶排序
 * @param arr 待排序数组
 */
function bottleSort(arr: number[]): void {
    if (arr.length === 0) return;

    for (let i = 0; i < arr.length; i++) {
        const n = arr[i];
        if (!(Number.isInteger(n) && n >= 0)) throw new Error("桶排序只支持非负整数");
    }

    const bitNum = Math.max(...arr).toString().length;
    for (let i = 0; i < bitNum; i++) {
        const buckets: number[][] = Array.from({ length: 10 }, () => []);
        for (let j = 0; j < arr.length; j++) {
            const index = Math.floor(arr[j] % Math.pow(10, i + 1) / Math.pow(10, i));
            buckets[index].push(arr[j]);
        }
        arr.splice(0, arr.length, ...buckets.flat());
    }
}

/**
 * 计数排序
 * @param arr 待排序数组
 */
function countingSort(arr: number[]): void {
    if (arr.length === 0) return;

    for (let i = 0; i < arr.length; i++) {
        if (!Number.isInteger(arr[i])) throw new Error("计数排序只支持整数");
    }

    const min = Math.min(...arr), max = Math.max(...arr);
    const offset = min;
    const length = max - min + 1;

    let array: number[] = Array.from({ length: length }, () => 0);
    for (let i = 0; i < arr.length; i++) {
        array[arr[i] - offset]++;
    }

    arr.splice(0, arr.length);
    for (let i = 0; i < length; i++) {
        arr.push(...Array.from({ length: array[i] }, () => i + offset));
    }
}

/**
 * 断言函数
 * @param condition 条件
 * @param msg 错误信息
 */
function assert(condition: boolean, msg?: string): void { if (!condition) { throw new Error(msg); } }

/**
 * 测试函数
 * @param sortFunc 待测试的排序函数
 */
function test(sortFunc: (arr: number[]) => void): void {

    /**
     * 判断两个数组是否值相等
     * @param a 数组a
     * @param b 数组b
     * @returns 是否值相等
     */
    function arrEqual(a: number[], b: number[]): boolean { return a.length === b.length && a.every((v, i) => v === b[i]); }

    // 基础排序测试
    let arr = [1, 10, 2, 9, 3, 8, 4, 7, 5, 6];
    sortFunc(arr);
    assert(arrEqual(arr, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]));

    arr = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1];
    sortFunc(arr);
    assert(arrEqual(arr, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]));

    // 测试已经排序好了的情况
    arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    sortFunc(arr);
    assert(arrEqual(arr, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]));

    // 测试重复元素情况
    arr = [3, 1, 2, 2, 3, 1, 4, 1, 2];
    sortFunc(arr);
    assert(arrEqual(arr, [1, 1, 1, 2, 2, 2, 3, 3, 4]));

    // 测试单个元素情况
    arr = [1];
    sortFunc(arr);
    assert(arrEqual(arr, [1]));

    // 测试空数组
    arr = [];
    sortFunc(arr);
    assert(arrEqual(arr, []));
}

test(insertSort);
test(shellSort);
test(selectSort);
test(bubbleSort);
test(quickSort);
test(mergeSort);
test(heapSort);
test(bottleSort);
test(countingSort);

console.log('All tests passed!');
